include/roboptim/core/detail/structured-input.hh

```diff
@@ -0,0 +1,114 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// roboptim is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.
+
+#ifndef ROBOPTIM_CORE_DETAIL_STRUCTURED_INPUT_HH
+# define ROBOPTIM_CORE_DETAIL_STRUCTURED_INPUT_HH
+# include <boost/mpl/assert.hpp>
+# include <roboptim/core/fwd.hh>
+# include <boost/type_traits.hpp>
+
+namespace roboptim
+{
+  namespace detail
+  {
+    /// \brief Gives access to a std::vector of std::pair<size_t, size_t> representing blocks of input
+    /// Those blocks are needed even if the function is not differentiable,
+    /// making it easier for us to define it in a exterior class to then use it transparently.
+    struct BlockProvider
+    {
+      /// \brief stores the blocks defined by the function as a pair of integers.
+      /// The first element is the first index of the block, the second element
+      /// repersetn the total size of the block.
+      std::vector<std::pair<size_t, size_t> > blocks;
+    };
+
+    /// \brief This class gives access to the getJacobianBlock() method, which implementation
+    /// is to be spacialized according to the type of matrix handled by the function.
+    template<typename FuncType, typename MatType>
+    struct StructuredInputJacobianInternal : public BlockProvider
+    {};
+
+    /// \brief DenseMatrix specialization of the getJacobianBlock() method provider.
+    template<typename FuncType>
+    struct StructuredInputJacobianInternal<FuncType, roboptim::EigenMatrixDense> : public BlockProvider
+    {
+      /// \brief return type of the getJacobianBlock() method
+      typedef typename FuncType::jacobian_ref JacBlock;
+      /// \brief input type of the jacobian given by the user
+      typedef typename FuncType::jacobian_ref InputJacBlock;
+
+      /// \brief retrieve a specific block of the jacobian given as an input
+      /// \param jacobian the jacobian to operate upon
+      /// \param blockInd the index of the block to be retrieved
+      JacBlock getJacobianBlock(InputJacBlock jacobian, size_t blockInd) const;
+    };
+
+    /// \brief DenseMatrix specialization of the getJacobianBlock() method provider.
+    template<typename FuncType>
+    struct StructuredInputJacobianInternal<FuncType, roboptim::EigenMatrixSparse> : public BlockProvider
+    {
+      /// \brief return type of the getJacobianBlock() method
+      typedef Eigen::Block<Eigen::SparseMatrix<double>, -1, -1, 0> JacBlock;
+      /// \brief input type of the jacobian given by the user
+      typedef Eigen::SparseMatrix<double>& InputJacBlock;
```

Comment:
I think we should get the actual RobOptim Jacobian type here rather than assuming it's `Eigen::SparseMatrix<double>` (because that's [actually false](https://github.com/roboptim/roboptim-core/blob/master/include/roboptim/core/function.hh#L656) if the storage order changes). Also, using Eigen's enum is usually clearer than `-1/0` since it's quite difficult to know for each Eigen type if this refers to the storage order, the size, the stride, etc.